gqcp/include/Mathematical/Representation/Tensor.hpp

```diff
@@ -287,37 +291,124 @@ class Tensor:
+        boost::replace_all(contraction_string, "-", " ");
+        boost::replace_all(contraction_string, ",", " ");
+
+        // split the stringstream in the necessary components, 3 in most cases.
+        std::vector<std::string> segment_list;
+        boost::split(segment_list, contraction_string, boost::is_any_of(" "));
+
+        // If a contraction over all indices is done, only 2 sets of labels are saved in the vector. The last set of labels is an empty string, and is added manually.
+        if (segment_list.size() == 2) {
+            segment_list.push_back("");
+        }
+
+        // The following code can be used to calculate template parameter `N` at runtime. At the moment however, this is not used.
+        // Determine number of axes to contract over
+        //
+        // std::vector<char> segment_1(segment_list[0].begin(), segment_list[0].end());
+        // std::vector<char> segment_2(segment_list[1].begin(), segment_list[1].end());
+        // std::vector<char> intersection;
+        //
+        // std::set_intersection(segment_1.begin(), segment_1.end(),
+        //                       segment_2.begin(), segment_2.end(), std::back_inserter(intersection));
+        //
+        // const int Z = intersection.size();
+
+        return this->einsum<N>(rhs, segment_list[0], segment_list[1], segment_list[2]);
+    }
+
+
+    /**
+     *  Contract this tensor with a matrix, using a NumPy 'einsum'-like API.
+     * 
+     *  @param contraction_string   The string used to specify the wanted contraction, e.g. "ijkl,jk->il"
+     *  @param rhs                  The right-hand side of the contraction, a matrix in this case.
+     * 
+     *  @tparam N                   The number of axes that should be contracted over.
+     * 
+     *  @example T.einsum("ijkl,jk->il", M) will contract the j and k axes of the second tensor with those of the first tensor, resulting in a rank 2 tensor with axes i and l.
+     * 
+     *  @return The result of the tensor contraction.
+     */
+    template <int N, int LHSRank = Rank>
+    Tensor<Scalar, LHSRank + 2 - 2 * N> einsum(std::string contraction_string, const Matrix<Scalar> rhs) const {
+        // convert matrix to rank 2 tensor
+        const auto tensor_from_matrix = Tensor<Scalar, 2>(Eigen::TensorMap<Eigen::Tensor<const Scalar, 2>>(rhs.data(), rhs.rows(), rhs.cols()));
+        return this->einsum<N>(contraction_string, tensor_from_matrix);
+    }
+
+
+    /**
+     *  Convert the given tensor to a matrix with the specified dimensions
+     *
+     *  @param rows        the dimension of the rows of the output matrix
+     *  @param cols        the dimension of the columns of the output matrix
+     */
+    const GQCP::Matrix<Scalar> toMatrix(const size_t rows, const size_t cols) const {
+
+        const auto rank_two_tensor = *this;
+        return GQCP::Matrix<Scalar>(Eigen::Map<const Eigen::Matrix<Scalar, Eigen::Dynamic, Eigen::Dynamic>>(rank_two_tensor.data(), rows, cols));
     }
```

Comment:
I think you can clean up this API even more. Could you also add the appropriate `enable_if`?

```suggestion
    /**
     *  @return This rank-two tensor as a matrix.
     */
    const GQCP::Matrix<Scalar> asMatrix() const {

        const auto rows = this->dimension(0);
        const auto cols = this->dimension(1);

        return GQCP::Matrix<Scalar>(Eigen::Map<const Eigen::Matrix<Scalar, Eigen::Dynamic, Eigen::Dynamic>>(this->data(), rows, cols));
    }
```